src/control/ros/Conversions.cpp

```diff
@@ -0,0 +1,283 @@
+    splineProblem.addConstantConstraint(1, 2, _nextAcceleration);
+  }
+
+  const auto splineSegment = splineProblem.fit();
+  return splineSegment.getCoefficients()[0];
+}
+
+//=============================================================================
+void extractJointTrajectoryPoint(
+  const trajectory_msgs::JointTrajectory& _trajectory,
+  size_t _index, size_t _numDofs,
+  Eigen::VectorXd* _positions, bool _positionsRequired,
+  Eigen::VectorXd* _velocities, bool _velocitiesRequired,
+  Eigen::VectorXd* _accelerations, bool _accelerationsRequired)
+{
+  const auto& waypoint = _trajectory.points[_index];
+
+  try
+  {
+    checkVector("positions", waypoint.positions, _numDofs,
+      _positionsRequired, _positions);
+    checkVector("velocities", waypoint.velocities, _numDofs,
+      _velocitiesRequired, _velocities);
+    checkVector("accelerations", waypoint.accelerations, _numDofs,
+      _accelerationsRequired, _accelerations);
+  }
+  catch (const std::invalid_argument& e)
+  {
+    std::stringstream message;
+    message << "Waypoint " << _index << " is invalid: " << e.what();
+    throw std::invalid_argument(message.str());
+  }
+}
+
+// The rows of vector and matrix will be reordered using permutationMatrix
+void reorder(const Eigen::MatrixXd& permutationMatrix,
+  const Eigen::VectorXd& inVector,
+  const Eigen::MatrixXd& inMatrix,
+  Eigen::VectorXd* outVector,
+  Eigen::MatrixXd* outMatrix)
+{
+  *outVector = permutationMatrix*inVector;
+  *outMatrix = permutationMatrix*inMatrix;
+}
+
+// Convert the mapping to a permutation matrix for matrix operation
+void convertToPermutationMatrix(const std::map<size_t, size_t>& map,
+  Eigen::MatrixXd* permutationMatrix)
+{
+  *permutationMatrix = Eigen::MatrixXd::Zero(map.size(), map.size());
+  for(auto it = map.begin(); it != map.end(); ++it)
+  {
+    (*permutationMatrix)(it->second, it->first) = 1.0;
+  }
+}
+
+} // namespace
+
+//=============================================================================
+std::unique_ptr<SplineTrajectory> convertJointTrajectory(
+  const std::shared_ptr<MetaSkeletonStateSpace>& space,
+  const trajectory_msgs::JointTrajectory& jointTrajectory)
+{
+  if (!space)
+    throw std::invalid_argument{"StateSpace must be non-null."};
+
+  const auto numControlledJoints = space->getNumSubspaces();
+  if (jointTrajectory.joint_names.size() != numControlledJoints)
+  {
+    std::stringstream message;
+    message << "Incorrect number of joints: expected "
+        << numControlledJoints << ", got "
+        << jointTrajectory.joint_names.size() << ".";
+    throw std::invalid_argument{message.str()};
+  }
+
+  // Check that all joints are single DOF.
+  for (size_t i = 0; i < space->getNumSubspaces(); ++i)
+  {
+    auto n = space->getJointSpace(i)->getJoint()->getNumDofs();
+    if (n != 1)
+    {
+      std::stringstream message;
+      message << "Expected 1 dof. Joint " << i << " has " << n << " dofs.";
```

Comment:
Nit: It would be helpful to include `joint->getName()` and `joint->getType()` to `message`, e.g.:

> Only single-DOF joints are supported. Joint "my_joint1" (index: 1) is a SphericalJoint with 3 DOFs.